p1-staking-pool-dyn/src/lib.rs

```diff
@@ -192,25 +192,26 @@ impl Contract {
             to_transfer
         );
 
-        // since NEAR .transfer never fails, we better discount the unstaked here,
-        // update the vault to avoid allowing double-unstake if any other promise fails,
-        // also zero the rewards to block double-withdraw-cheddar
-        vault.staked = 0;
-        vault.rewards = 0;
-        self.save_vault(&aid, &vault);
+        // since NEAR transfer never fails, we must remove the vault here to clean up the
+        // storage and protect against double-unstake or double harvest (which would be
+        // possible  other promise will fail),
+        self.vaults.remove(&aid);
         self.total_stake -= to_transfer;
 
         // 1st promise is to transfer back all their NEAR, we assume near transfer does not fail
         assert!(to_transfer > 0);
-        Promise::new(aid.clone()).transfer(to_transfer);
-        // note: returning an "and/joint" promises is forbidden now
-        // and joining the 2 promises with "then" causes the "transfer"
-        // to become the main promise for the callback,
-        // so the success/failure of the mint-call can not be evaluated.
-        // Creating 2 promises (batch) works correctly
+        let p = Promise::new(aid.clone()).transfer(to_transfer);
```

Comment:
This is incorrect, the mere fact of invoking Promise::new schedules the promise to be executed.